File: Python/algorithmic_heights/14_3sum.py
# 3SUM
import sys


def convert_to_intlist(line):
    return tuple([int(item) for item in line.split()])


def read_arrays(reader, k):
    return [convert_to_intlist(next(reader)) for _ in range(k)]


def solve_2_sum(array, target, target_ix):
    negative_target_indices = {}
    for ix, item in enumerate(array[target_ix+1:], start=target_ix+1):
        if (jy := negative_target_indices.get(item)) is not None:
            return sorted((jy+1, ix+1, target_ix+1))

        negative_target_indices[target-item] = ix
    return None


def find_zero_sum_index_pairs(arrays):
    def solve_3sum(array):
        for ix, item in enumerate(array):
            if (indices := solve_2_sum(array, -item, ix)) is not None:
                return indices
        return (-1,)

    return [solve_3sum(array) for array in arrays]


def main():
    reader = sys.stdin
    k, _ = convert_to_intlist(next(reader))
    arrays = read_arrays(reader, k)
    result = find_zero_sum_index_pairs(arrays)
    for indices in result:
        print(' '.join(map(str, indices)))


if __name__ == '__main__':
    main()

File: Haskell/algorithmic_heights/14_3sum.hs
import Control.Monad (mapM_, replicateM)
import Data.IntMap (IntMap)
import qualified Data.IntMap as M
import Data.Maybe (listToMaybe, mapMaybe)


convertToIntList :: String -> [Int]
convertToIntList = map read . words


readParameters :: String -> (Int, Int)
readParameters line =
    case convertToIntList line of
        [a, b] -> (a, b)
        _ -> error "Malformed input."


readLists :: Int -> IO [[Int]]
readLists k = do
    rows <- replicateM k getLine
    return $ map convertToIntList rows


solve2Sum :: [Int] -> Int -> Int -> Maybe (Int, Int, Int)
solve2Sum xs target targetIx =
    let go negativeTargetIndices [] = Nothing
        go negativeTargetIndices ((item, ix) : rest) =
            case M.lookup item negativeTargetIndices of
                Nothing -> go (M.insert (target - item) ix negativeTargetIndices) rest
                Just jy -> Just (targetIx + 1, jy + 1, ix + 1)
    in go M.empty $ drop (targetIx + 1) $ zip xs [0..]


findZeroSumIndexTriples :: [[Int]] -> [Maybe (Int, Int, Int)]
findZeroSumIndexTriples lists =
    let solve xs = listToMaybe $ mapMaybe (\(item, ix) -> solve2Sum xs (-item) ix) $ zip xs [0..]
    in map solve lists


resultToString :: Maybe (Int, Int, Int) -> String
resultToString Nothing = "-1"
resultToString (Just (i, j, k)) = show i ++ " " ++ show j ++ " " ++ show k


main :: IO ()
main = do
    (k, _) <- readParameters <$> getLine
    lists <- readLists k
    let result = findZeroSumIndexTriples lists
    mapM_ (putStrLn . resultToString) result
